src/v/cluster/tx_gateway_frontend.cc

```diff
@@ -69,6 +69,32 @@ static auto with(
       });
 }
 
+template<typename Func>
+auto tx_gateway_frontend::with_stm(Func&& func) {
+    auto partition = _partition_manager.local().get(model::tx_manager_ntp);
+    if (!partition) {
+        vlog(
+          txlog.warn, "can't get partition by {} ntp", model::tx_manager_ntp);
+        return func(tx_errc::partition_not_found);
+    }
+
+    auto stm = partition->tm_stm();
+
+    if (!stm) {
+        vlog(
+          txlog.warn,
+          "can't get tm stm of the {}' partition",
+          model::tx_manager_ntp);
+        return func(tx_errc::stm_not_found);
+    }
+
+    if (!stm->gate().try_enter()) {
+        return func(tx_errc::unknown_server_error);
+    }
+
+    return func(stm).finally([stm] { stm->gate().leave(); });
+}
```

Comment:
nit: this looks safe. the preferred RAII verison would be

```
if gate.is_closed() return;
return with_gate(gate, func()...)
```

Reply:
changed